exporter/elasticsearchexporter/integrationtests/exporter_test.go

```diff
@@ -0,0 +1,164 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package integrationtests
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/tidwall/gjson"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/plog/plogotlp"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+func TestExporter(t *testing.T) {
+	for _, tc := range []struct {
+		name             string
+		restartCollector bool
+		mockESFailure    bool
+	}{
+		{name: "basic"},
+		{name: "es_intermittent_failure", mockESFailure: true},
+		/* Below tests should be enabled after #30792 is fixed
+		{name: "collector_restarts", restartCollector: true},
+		{name: "collector_restart_with_es_intermittent_failure", mockESFailure: true, restartCollector: true},
+		*/
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			runner(t, tc.restartCollector, tc.mockESFailure)
+		})
+	}
+}
+
+func runner(t *testing.T, restartCollector, mockESFailure bool) {
+	t.Helper()
+
+	cfg, err := loadConfig()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	mockES, err := newMockESClient(t, cfg.Debug)
+	require.NoError(t, err)
+
+	collector, err := newTestCollector(t, cfg, mockES.ServerURL)
+	require.NoError(t, err)
+	t.Cleanup(collector.Shutdown)
+
+	var g errgroup.Group
+	cancelRun := runTestCollectorWithWait(ctx, t, collector, &g)
+
+	if mockESFailure {
+		mockES.SetReturnStatusCode(http.StatusServiceUnavailable)
+		if !restartCollector {
+			// If restartCollector is not set then recover ES service
+			// after a few failed bulk requests otherwise ensure that
+			// both collector and ES are unavailable for the same
+			// duration considering worst case scenario.
+			time.AfterFunc(10*time.Second, func() {
```

Comment:
Depending on timing in tests makes me nervous as it could lead to flaky tests 😅.  Can we know/calculate how many bulk requests, in all, the collector will make to ES in this test? If so, maybe we can come up with a way to set the ES return status code to 200 OK after, say, half that number of bulk requests have failed?

Reply:
I share your fear of time-based flaky tests 😅. The current approach is not non-deterministic. The test configures a flush interval of 1 second ([ref](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3ee127639bdf2c94fd38769cc3a781171fed82a6/exporter/elasticsearchexporter/integrationtests/collector.go#L171-L172)) and restarts ES after 10 seconds. Note that ES is down when we start the test and we recover it after 10 seconds. So, the only thing that we need to ensure is that there are a few _bulk requests before 10 seconds are up and the flush interval _should_ take care of this.

> Can we know/calculate how many bulk requests, in all, the collector will make to ES in this test? If so, maybe we can come up with a way to set the ES return status code to 200 OK after, say, half that number of bulk requests have failed?

That is a good idea, it would bloat the mock ES a bit but it should work. Calculating an exact number of bulk requests might require exposing internal implementation details for the exporter and the underlying indexer but we can probably just fail the bulk indexer a fixed number of times instead. Not sure if it would be any better than what we have now though, WDYT?

Reply:
Okay, I understand why this test is deterministic, thanks.  However, there is an [implicit dependency](https://connascence.io/value.html) here between the flush interval and the wait duration values.  Ideally, we could make the wait time here a multiple of the flush interval to make that dependency explicit but if that's adding too much complexity, let's just comment here explaining that the wait duration needs to be a multiple of the flush interval value defined in the test collector's configuration.

Reply:
Good point. I have included flush interval in the template and updated the code to derive the ES downtime based on the flush interval: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31848/commits/81c2fb853d3c0c1e7aebe7a347cae8bd2ad11cf8